### PR TITLE
add initial message endpoint

### PR DESCRIPTION
**Pull request by brngylni**
This PR adds the initial message endpoint. #5

### CODE REVIEW

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:1]: Please remove the empty line from the top of the file.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:10]: Please leave two lines of difference between each segregated block definition.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:28]: Please read this from a configuration file rather than hardcoding it into the code itself.

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/util.py:47]: Is it intentional to have the identical code block for the function definition of `verify_signature` in the `parser` module? If not, please remove it from either of the locations.

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/util.py:51]: GitHub is a proper noun.

```suggestion
    """Get the Fedora Account System Username of the given GitHub username
```

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (gridhead) [inline webhook_to_fedora_messaging/endpoints/util.py:54]: GitHub is a proper noun.

```suggestion
            username: GitHub Username"""
```

**Review by gridhead — CHANGES_REQUESTED** [review]
@brngylni Please squash the commits and every other changes required on the codebase can be done after the merge.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/parser/parser.py:28]: Will change here after the database models are changed

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:6]: You need to add it as a dependency, with `poetry add webhook-to-fedora-messaging-messages`.

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/message.py:30]: This is the old way of make SQLAlchemy queries. The new style is:
```suggestion
from sqlalchemy import select
from sqlalchemy.exc import NoResultFound
try:
    service = db.session.scalar(select(Service).where(Service.uuid == service_uuid)).one()
except NoResultFound:
    return not_found()
```

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/message.py:27]: You don't need this, you can just use `db.session`

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/message.py:32]: With the try/except block above, you don't need this.

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/message.py:34]: You don't need to pass `request` around, it's a global object.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:11]: Same comment about passing `request` around, valid for the whole file. Just use `flask.request`.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:13]: The function should raise `ValueError` if the `type` isn't handled. You can do that at the end of the function.

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/message.py:34]: Also, if as suggested below this function can raise `ValueError` on unknown service types, this should be caught here to return something appropriate (a 4xx response type).

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:18]: The first line (the summary) should be right after the quote marks:
```suggestion
"""Convert Flask request objects into desired FedMsg format.
```

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:23]: Those must be de-indented of 4 spaces, to be aligned with the quote marks.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:29]: It's usually clearer to handle the error cases first. For example, instead of:
```python
if condition 1:
    if condition 2:
        return do_something()
    return "2 not matched"
return "1 not matched"
```
it is usually clearer to do:
```python
if not condition1:
    return "1 not matched"
if not condition2:
    return "2 not matched"
return do_something()
```
This avoid indenting too much, and puts the response of the error cases closer to the condition.
The reasoning is: deal with the error cases first, and then do your normal path.

As a result, here I would do something like:
```python
if secret and not verify_signature(...):
    return None
topic = ...
agent = ...
return GithubMessageV1(...)
```

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:33]: If the signature didn't match, we should raise an exception. A custom exception (to put in `webhook_to_fedora_messaging.exceptions`) is best, but a ValueError could also do.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:43]: Actually you can just do `dict(headers)`. No need for this function.

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/util.py:7]: This is equivalent to doing `return {'message', 'Not Found'}, 404`:
- if you return a dict from a view, Flask will automatically convert it to JSON and set the Content-Type header
- if you return a tuple from a view with a number as the second element, Flask will use it as the return code.

Please look at [this Flask doc about error handling in API servers](https://flask.palletsprojects.com/en/2.3.x/errorhandling/#returning-api-errors-as-json)

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/util.py:11]: This is equivalent to just returning `data`, no need for this function.

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/util.py:15]: This is equivalent to returning `{'message': 'Bad Request'}`, 400. Please see the doc above to setup a JSON-aware error handler that will allow us to just raise exceptions or use `flask.abort()`.

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/util.py:19]: This is equivalent to returning `data, 201`, no need for this function

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/util.py:23]: Same thing here.

**Review by abompard — CHANGES_REQUESTED** [review]

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/util.py:47]: I'll remove it from util file

File: webhook_to_fedora_messaging/endpoints/util.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/util.py:7]: I'm going to fix it here https://github.com/fedora-infra/webhook-to-fedora-messaging/pull/32/files

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (gridhead) [reply webhook_to_fedora_messaging/endpoints/message.py]: @abompard It seems that @brngylni has a separate PR for the `util` module that is intended to fill the shoes here. As such the said `util` module does not exist in the targetted branch and hence, the inclusion of it would assume that a new file is created. @brngylni intends to merge #32 first i.e. the branch that has the `util` module shared by all of these endpoint modules. As such, I suggest @brngylni either include the `create_event` and `query_fasjson` in that PR or wait for that PR to be merged first before you rebase atop the newly-changed targetted branch with those additions to the `util` module in this PR.

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/message.py:24]: excuse me which try except?

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/message.py:13]: This may conflict with other branches I guess. We might need another pr to switch this to the one you supposed. Since the other branches are merged.

**Review by gridhead — APPROVED** [review]
I am not dragging this any longer than it has to be. Other required changes can be made to the codebase after it is merged.

Merging now.

File: webhook_to_fedora_messaging/config/defaults.py
Comment (abompard) [reply webhook_to_fedora_messaging/config/defaults.py]: You still need to set the development value in `devel/ansible/roles/dev/files/config.toml`

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [reply webhook_to_fedora_messaging/endpoints/message.py:13]: I'm not sure I understand. Which branches will it conflict with?

Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/message.py:13]: I have the code you suggested on local but after a meeting with @gridhead we thought the current one might be better. So whichever you guys decide I can push.

**Comment by abompard** [comment]
Please rebase this branch, I'll review it afterwards.

**Comment by brngylni** [comment]
> Please rebase this branch, I'll review it afterwards.

Rebased it

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:20]: This args list does not match the actual arguments.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:31]: It's preferable to use string formatting rather than concatenation, for performance reasons:

```suggestion
    topic = f"github.{headers['X-Github-Event']}"
```

File: webhook_to_fedora_messaging/exceptions/exceptions.py
Comment (abompard) [inline webhook_to_fedora_messaging/exceptions/exceptions.py:1]: There is already an `exceptions.py` file in the main directory, please use it.

File: webhook_to_fedora_messaging/endpoints/parser/parser.py
Comment (abompard) [inline webhook_to_fedora_messaging/endpoints/parser/parser.py:17]: Please put this Github-specific code into a `github.py` module (also put the `verify_signature()` function there as it is Github-specific as well).

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/message.py:38]: I'm not sure if messages have an id attribute in our case. Since we don't log the messages, I thought we don't have that

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (brngylni) [reply webhook_to_fedora_messaging/endpoints/message.py:40]: I think we can remove the else block since we already use a try-except block above

File: webhook_to_fedora_messaging/endpoints/message.py
Comment (abompard) [reply webhook_to_fedora_messaging/endpoints/message.py]: Yeah this message object is a Fedora Messaging Message instance, so it does have an ID. It's not a database object.

**Review by abompard — APPROVED** [review]
Looks good, thanks!

**Pull request merged**
